### PR TITLE
Fix tests with cross binary arguments

### PR DESCRIPTION
**Pull request by dcbaker**
This series contains a couple of bug fixes and some tests. It has the tests that should have come with #5971. It also has a fix for #5982 (which is what I set out to fix), as well as a fix for #5984, which I discovered while writing the tests for 5982.

### CODE REVIEW

File: run_unittests.py [hunk 1]
```diff
@@ -6535,6 +6482,130 @@ class CrossFileTests(BasePlatformTests):
...
+            needs_exe_wrapper = {}
+            {}
+
+            [host_machine]
+            system = 'linux'
+            cpu_family = 'x86'
+            cpu = 'i686'
+            endian = 'little'
+            """.format(
+                cc, needs_exe_wrapper,
```
Comment (ignatenkobrain) [inline run_unittests.py]: __[Flake8]__ [E126] continuation line over-indented for hanging indent ([link](https://sider.review/gh/repos/19784232/pulls/5986#issue-4635090))

<sub>posted by Sider</sub>

Comment (dcbaker) [reply run_unittests.py]: this really annoys me because it's wrong, that is the correct indent.

**Comment by dcbaker** [comment]
@jpakkane, It would be great for me if this could go into the 0.52.0 release.

**Comment by jpakkane** [comment]
I missed this. Maybe for the point release then?

**Comment by dcbaker** [comment]
That would be great if possible. This bug is responsible for the last set of test failures in the meson-for-windows branch of mesa, and there isn't really a good way to work around it.

**Comment by dcbaker** [comment]
I've changed my mind about getting this into 0.52.1. I can actually work around this easily in the test scripts by catching an OSError and checking to see if it's a NOEXEC error. I think before we implement this I want to fix  #4427 as well, which will round out the whole situation and let scripts always work correctly.

**Comment by dcbaker** [comment]
I have tested the pass exe_wrapper with mesa, but I still need to write a proper test case for it

**Comment by jpakkane** [comment]
Is this ready now in your opinion?

**Comment by dcbaker** [comment]
@jpakkane, I think it is. assuming that my rebase went well.

**Comment by jpakkane** [comment]
The only issue I came up with was tests that take the executable as an argument but don't actually try to run it. For example there may be a test that checks that the executable has certain symbols, or any other form of binary inspection. Those should still keep running, but unfortunately there is no way to tell from the outside whether a test script will attempt to run the program or not...

**Comment by dcbaker** [comment]
Yeah, that's true. Hummm. we have some of those kind of tests in mesa. Would it be better to drop that patch and add some documentation to the effect of "if you pass executables to scripts and need and exe_wapper but don't have on you should return 77"?

**Comment by jpakkane** [comment]
Something like that seems reasonable. We can fix it later if someone comes up with a good way to tackle this problem.

**Comment by dcbaker** [comment]
@jpakkane, here's a blast from the past I actually got successfully rebased this time.

I've check the behavior, by the final patch it set MESON_EXE_WRAPPER if you have one, and otherwise you're on your own to skip the tests by catching the NOEXEC error.

File: docs/markdown/snippets/exe_wrapper_for_cross_built_tests.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+## Test scripts are given the exe wrapper if needed
+
+Meson will now set the `MESON_EXE_WRAPPER` as the `' '` joined command of the
+exe wrapper. This allow wrapper scripts to run test binaries, instead of just
+skipping.
```
Comment (jpakkane) [inline docs/markdown/snippets/exe_wrapper_for_cross_built_tests.md:5]: This should probably be using `shlex.join` instead to make those Windows file names with spaces in them work.

Comment (dcbaker) [reply docs/markdown/snippets/exe_wrapper_for_cross_built_tests.md]: Updated to use `mesonlib.join_args` (which is more correct for windows than `shlex.join`)

**Pull request merged**